File: py/communicate_test.py
from websockets.sync.client import connect
import json
import time
# import sys
# sys.stdout.write("Hello")
# sys.stdout.flush()
# while True:
    # print("Hello")
    # sys.stdin.flush()
    # lines = sys.stdin.readlines()
    # if (lines == None or len(lines) < 1):
        # time.sleep(1)
        # print(len(lines))
    #     continue
    # line = lines[0]
    # line = input()
    # print(line)
    # try:
        # data = json.loads(line)
    #     if (data['msg'] == "Ping!"):
    #         sys.stdout.write(json.dumps({
    #             'msg': "Pong!"
    #         }))
    #         sys.stdout.flush()
    # except SystemExit:
    #     raise
    # except:
    #     pass

with connect("ws://localhost:9910/ws/") as websocket:
    while True:
        try:
            data = json.loads(websocket.recv())
            if data["msg"] == "ping":
                print("Ping Recieved")
                websocket.send(json.dumps({
                    'msg': "Pong!"
                }))
            elif data["msg"] == "quit":
                exit(0)
            else:
                print(data)
        except SystemExit:
            raise
        except:
            pass

File: js/controller/communicate_test.js
require("./.pnp.cjs").setup()
let child_process = require("child_process")
let express = require("express")
let ews = require("express-ws")
let http = require("http")

let app = express()
let wsRouter = express.Router()
app.use(wsRouter)
let server = http.createServer(app)
let instance = ews(app, server)

let rl = require('readline').createInterface({
    input: process.stdin,
})

let python = child_process.spawn("python3", ["../../py/communicate_test.py"], {
    shell: true,
    // stdio: 'ignore'
    stdio: ["ignore", "inherit", "inherit"]
})

wsRouter.ws("/ws", async ws => {
    ws.on("message", (_msg) => {
        console.log(_msg)
    })
})

// let stream = require("stream")
let inv = setInterval(() => {
    console.log("Sending Ping")
    instance.getWss().clients.forEach(client => {
        client.send(JSON.stringify({
            msg: "Ping!"
        }))
    })
    // console.log(python.stdin.write(JSON.stringify({
    //     msg: "Ping!"
    // })))
    // python.stdin.uncork()
    // let r = new stream.PassThrough()
    // r.pipe(python.stdin)
    // r.end(JSON.stringify({
    //     msg: "Ping!"
    // })+'\n')
    // r.end(,)
    // python.stdin.write("H")
    // python.stdin.end()

}, 5000)
/* python.stdout.setEncoding("utf-8").on("data", (_data) => {
    try {
        let data = JSON.parse(_data)

        if (data.msg == "Pong!") {
            console.log("Received Pong")
        } else {
            console.log(_data)
        }
    } catch {
        if (_data == " " || _data == "\n") {
            return
        }
        console.log(_data)
    }
}) */

server.listen(9910)

let closing = false

let exit = (c = 0) => {
    closing = true
    clearInterval(inv)
    rl.close()
    for (let client of instance.getWss().clients) {
        client.send(JSON.stringify({
            msg: "quit"
        }))
    }
    setTimeout(() => {
        process.exit(c)
    }, 2000)
}

python.on("close", () => {
    if (closing) {
        return
    }
    console.error("Python has been closed")
    exit(2)
})

rl.on("line", (i) => {
    if (i == 'q') {
        exit(0)
    }
})
